File: tugas-7-20251008/python/Soal3.py
from datetime import datetime

class JenisKendaraan:
    MOTOR = "MOTOR"
    MOBIL = "MOBIL"

class Status:
    REGULAR = "REGULAR"
    MENGINAP = "MENGINAP"

class Time:
    def __init__(self, hour=0, minute=0, second=0):
        self.__hour = hour
        self.__minute = minute
        self.__second = second

    def get_hour(self):
        return self.__hour

    def get_minute(self):
        return self.__minute

    def get_second(self):
        return self.__second

    def __str__(self):
        return f"{self.__hour:02d}:{self.__minute:02d}:{self.__second:02d}"

class Date:
    def __init__(self, year=0, month=0, day=0):
        self.__year = year
        self.__month = month
        self.__day = day

    def get_year(self):
        return self.__year

    def get_month(self):
        return self.__month

    def get_day(self):
        return self.__day

    def __str__(self):
        return f"{self.__year:04d}/{self.__month:02d}/{self.__day:02d}"

class Waktu:
    def __init__(self, time=None, date=None):
        self.__time = time
        self.__date = date

    def get_time(self):
        return self.__time

    def get_date(self):
        return self.__date

class ParkedVehicle:
    def __init__(self, no_kendaraan, kendaraan, status, datang, pulang):
        self.__no_kendaraan = no_kendaraan
        self.__kendaraan = kendaraan
        self.__status = status
        self.__datang = datang
        self.__pulang = pulang

    def get_no_kendaraan(self):
        return self.__no_kendaraan

    def get_kendaraan(self):
        return self.__kendaraan

    def get_status(self):
        return self.__status

    def get_datang(self):
        return self.__datang

    def get_pulang(self):
        return self.__pulang

class ParkedVehicleHelper:
    @staticmethod
    def get_pay(vehicle, total_jam, total_hari):
        if vehicle.get_status() == Status.MENGINAP:
            multiplier = 15000 if vehicle.get_kendaraan() == JenisKendaraan.MOTOR else 25000
            return multiplier * max(total_hari, 1)
        else:
            multiplier = 2000 if vehicle.get_kendaraan() == JenisKendaraan.MOTOR else 3000
            return multiplier * total_jam

def parse_time(s):
    h, m, s = map(int, s.split(":"))
    return Time(h, m, s)

def parse_date(s):
    y, m, d = map(int, s.split("/"))
    return Date(y, m, d)

def calculate_total_time(datang, pulang):
    datang_dt = datetime(
        datang.get_date().get_year(),
        datang.get_date().get_month(),
        datang.get_date().get_day(),
        datang.get_time().get_hour(),
        datang.get_time().get_minute(),
        datang.get_time().get_second()
    )
    pulang_dt = datetime(
        pulang.get_date().get_year(),
        pulang.get_date().get_month(),
        pulang.get_date().get_day(),
        pulang.get_time().get_hour(),
        pulang.get_time().get_minute(),
        pulang.get_time().get_second()
    )

    diff = pulang_dt - datang_dt
    total_detik = diff.total_seconds()
    if total_detik < 0:
        total_detik += 86400

    total_hari = total_detik // 86400 
    sisa_detik = total_detik % 86400
    total_jam = sisa_detik // 3600
    if sisa_detik % 3600 > 0:
        total_jam += 1
    total_jam += total_hari * 24
    if total_hari >= 1 and sisa_detik == 0:
        total_jam = int(total_hari * 24)

    return int(total_hari), int(total_jam)

def print_vehicle_summary(pv, total_hari, total_jam, pay):
    print("\n--- RINCIAN KENDARAAN ---")
    print("Nomor Kendaraan :", pv.get_no_kendaraan())
    print("Jenis Kendaraan :", pv.get_kendaraan())
    print("Status Parkir   :", pv.get_status())
    print("Tanggal Datang  :", pv.get_datang().get_date())
    print("Tanggal Pulang  :", pv.get_pulang().get_date())
    print("Jam Datang      :", pv.get_datang().get_time())
    print("Jam Pulang      :", pv.get_pulang().get_time())
    print("Total Hari      :", total_hari)
    print("Total Jam       :", total_jam)
    print("Biaya           : Rp", pay)

def main():
    parked_vehicles = []
    total_pay = 0

    n = int(input("Masukkan jumlah kendaraan: "))
    for i in range(n):
        print(f"\n--- Data Kendaraan ke-{i+1} ---")
        no = input("Nomor Kendaraan: ")
        jenis = int(input("Jenis Kendaraan (0: Motor, 1: Mobil): "))
        stat = int(input("Status Parkir (0: Regular, 1: Menginap): "))

        print("\nWaktu Kedatangan")
        datang_tgl = input("Tanggal (yyyy/MM/dd): ")
        datang_jam = input("Jam (HH:mm:ss): ")

        print("\nWaktu Kepulangan")
        pulang_tgl = input("Tanggal (yyyy/MM/dd): ")
        pulang_jam = input("Jam (HH:mm:ss): ")

        datang = Waktu(parse_time(datang_jam), parse_date(datang_tgl))
        pulang = Waktu(parse_time(pulang_jam), parse_date(pulang_tgl))

        kendaraan = JenisKendaraan.MOTOR if jenis == 0 else JenisKendaraan.MOBIL
        status = Status.REGULAR if stat == 0 else Status.MENGINAP

        pv = ParkedVehicle(no, kendaraan, status, datang, pulang)
        parked_vehicles.append(pv)

    for pv in parked_vehicles:
        total_hari, total_jam = calculate_total_time(pv.get_datang(), pv.get_pulang())
        pay = ParkedVehicleHelper.get_pay(pv, total_jam, total_hari)
        total_pay += pay

        print_vehicle_summary(pv, total_hari, total_jam, pay)

    print("\nTotal Bayar Keseluruhan :", total_pay)

if __name__ == "__main__":
    main()

File: tugas-7-20251008/cpp/Soal3.cpp
#include <iostream>
#include <iomanip>
#include <string>
#include <sstream>
#include <cmath>
using namespace std;

// === ENUMS ===
enum JenisKendaraan { MOTOR, MOBIL };
enum Status { REGULAR, MENGINAP };

// === CLASS TIME ===
class Time {
private:
    int hour, minute, second;

public:
    Time() : hour(0), minute(0), second(0) {}
    Time(int h, int m, int s) : hour(h), minute(m), second(s) {}

    int getHour() const { return hour; }
    int getMinute() const { return minute; }
    int getSecond() const { return second; }

    void setHour(int h) { hour = h; }
    void setMinute(int m) { minute = m; }
    void setSecond(int s) { second = s; }

    string toString() const {
        ostringstream oss;
        oss << setw(2) << setfill('0') << hour << ":"
            << setw(2) << minute << ":"
            << setw(2) << second;
        return oss.str();
    }
};

// === CLASS DATE ===
class Date {
private:
    int year, month, day;

public:
    Date() : year(0), month(0), day(0) {}
    Date(int y, int m, int d) : year(y), month(m), day(d) {}

    int getYear() const { return year; }
    int getMonth() const { return month; }
    int getDay() const { return day; }

    void setYear(int y) { year = y; }
    void setMonth(int m) { month = m; }
    void setDay(int d) { day = d; }

    string toString() const {
        ostringstream oss;
        oss << setw(4) << setfill('0') << year << "/"
            << setw(2) << month << "/"
            << setw(2) << day;
        return oss.str();
    }
};

// === CLASS WAKTU ===
class Waktu {
private:
    Time time;
    Date date;

public:
    Waktu() {}
    Waktu(Time t, Date d) : time(t), date(d) {}

    Time getTime() const { return time; }
    Date getDate() const { return date; }

    void setTime(Time t) { time = t; }
    void setDate(Date d) { date = d; }
};

// === CLASS PARKED VEHICLE ===
class ParkedVehicle {
private:
    string noKendaraan;
    JenisKendaraan kendaraan;
    Status status;
    Waktu datang, pulang;

public:
    ParkedVehicle() {} // === PERUBAHAN: tambahkan konstruktor default
    ParkedVehicle(string no, JenisKendaraan k, Status s, Waktu d, Waktu p)
        : noKendaraan(no), kendaraan(k), status(s), datang(d), pulang(p) {}

    string getNoKendaraan() const { return noKendaraan; }
    JenisKendaraan getKendaraan() const { return kendaraan; }
    Status getStatus() const { return status; }
    Waktu getDatang() const { return datang; }
    Waktu getPulang() const { return pulang; }

    void setData(string no, JenisKendaraan k, Status s, Waktu d, Waktu p) {
        noKendaraan = no;
        kendaraan = k;
        status = s;
        datang = d;
        pulang = p;
    }
};

// === HELPER: HITUNG SELISIH WAKTU ===
class WaktuHelper {
public:
    static void calculateTotalTime(const Waktu &datang, const Waktu &pulang, int &totalHari, int &totalJam) {
        tm datang_tm = {};
        datang_tm.tm_year = datang.getDate().getYear() - 1900;
        datang_tm.tm_mon = datang.getDate().getMonth() - 1;
        datang_tm.tm_mday = datang.getDate().getDay();
        datang_tm.tm_hour = datang.getTime().getHour();
        datang_tm.tm_min = datang.getTime().getMinute();
        datang_tm.tm_sec = datang.getTime().getSecond();

        tm pulang_tm = {};
        pulang_tm.tm_year = pulang.getDate().getYear() - 1900;
        pulang_tm.tm_mon = pulang.getDate().getMonth() - 1;
        pulang_tm.tm_mday = pulang.getDate().getDay();
        pulang_tm.tm_hour = pulang.getTime().getHour();
        pulang_tm.tm_min = pulang.getTime().getMinute();
        pulang_tm.tm_sec = pulang.getTime().getSecond();

        time_t datang_time = mktime(&datang_tm);
        time_t pulang_time = mktime(&pulang_tm);

        double diff_seconds = difftime(pulang_time, datang_time);
        if (diff_seconds < 0) diff_seconds += 86400;

        totalHari = diff_seconds / 86400;
        double sisa = fmod(diff_seconds, 86400);
        totalJam = sisa / 3600;
        if (fmod(sisa, 3600) > 0) totalJam += 1;
        totalJam += totalHari * 24;
    }
};

// === HELPER: HITUNG BIAYA ===
class ParkedVehicleHelper {
public:
    static int getPay(const ParkedVehicle &v, int totalJam, int totalHari) {
        if (v.getStatus() == MENGINAP) {
            int rate = (v.getKendaraan() == MOTOR) ? 15000 : 25000;
            return rate * max(totalHari, 1);
        } else {
            int rate = (v.getKendaraan() == MOTOR) ? 2000 : 3000;
            return rate * totalJam;
        }
    }
};

// === PARSER ===
Time parseTime(const string &s) {
    int h, m, sec;
    char sep;
    istringstream iss(s);
    iss >> h >> sep >> m >> sep >> sec;
    return Time(h, m, sec);
}

Date parseDate(const string &s) {
    int y, m, d;
    char sep;
    istringstream iss(s);
    iss >> y >> sep >> m >> sep >> d;
    return Date(y, m, d);
}

// === OUTPUT ===
void printVehicleSummary(const ParkedVehicle &pv, int totalHari, int totalJam, int pay) {
    cout << "\n--- RINCIAN KENDARAAN ---\n";
    cout << "Nomor Kendaraan : " << pv.getNoKendaraan() << endl;
    cout << "Jenis Kendaraan : " << (pv.getKendaraan() == MOTOR ? "Motor" : "Mobil") << endl;
    cout << "Status Parkir   : " << (pv.getStatus() == REGULAR ? "Regular" : "Menginap") << endl;
    cout << "Tanggal Datang  : " << pv.getDatang().getDate().toString() << endl;
    cout << "Jam Datang      : " << pv.getDatang().getTime().toString() << endl;
    cout << "Tanggal Pulang  : " << pv.getPulang().getDate().toString() << endl;
    cout << "Jam Pulang      : " << pv.getPulang().getTime().toString() << endl;
    cout << "Total Hari      : " << totalHari << endl;
    cout << "Total Jam       : " << totalJam << endl;
    cout << "Biaya           : Rp " << pay << endl;
}

// === CLASS ARRAY KENDARAAN ===
class LarikKendaraan {
private:
    ParkedVehicle kendaraan[10];
    int n;

public:
    LarikKendaraan() : n(0) {}

    void tambahData(const ParkedVehicle &pv) {
        kendaraan[n++] = pv;
    }

    void tampilData() {
        int totalBayar = 0;
        for (int i = 0; i < n; i++) {
            int totalHari, totalJam;
            WaktuHelper::calculateTotalTime(kendaraan[i].getDatang(), kendaraan[i].getPulang(), totalHari, totalJam);
            int pay = ParkedVehicleHelper::getPay(kendaraan[i], totalJam, totalHari);
            totalBayar += pay;
            printVehicleSummary(kendaraan[i], totalHari, totalJam, pay);
        }
        cout << "\nTotal Bayar Keseluruhan : Rp " << totalBayar << endl;
    }
};

// === MAIN ===
int main() {
    int n;
    cout << "Masukkan jumlah kendaraan: ";
    cin >> n;
    cin.ignore();

    LarikKendaraan daftar; // === PERUBAHAN: ganti vector jadi objek array

    for (int i = 0; i < n; i++) {
        cout << "\n--- Data Kendaraan ke-" << i + 1 << " ---\n";
        string no, datangTgl, datangJam, pulangTgl, pulangJam;
        int jenis, stat;

        cout << "Nomor Kendaraan: ";
        getline(cin, no);
        cout << "Jenis Kendaraan (0: Motor, 1: Mobil): ";
        cin >> jenis;
        cout << "Status Parkir (0: Regular, 1: Menginap): ";
        cin >> stat;
        cin.ignore();

        cout << "\nWaktu Kedatangan\n";
        cout << "Tanggal (yyyy/MM/dd): ";
        getline(cin, datangTgl);
        cout << "Jam (HH:mm:ss): ";
        getline(cin, datangJam);

        cout << "\nWaktu Kepulangan\n";
        cout << "Tanggal (yyyy/MM/dd): ";
        getline(cin, pulangTgl);
        cout << "Jam (HH:mm:ss): ";
        getline(cin, pulangJam);

        Waktu datang(parseTime(datangJam), parseDate(datangTgl));
        Waktu pulang(parseTime(pulangJam), parseDate(pulangTgl));
        JenisKendaraan jk = (jenis == 0) ? MOTOR : MOBIL;
        Status st = (stat == 0) ? REGULAR : MENGINAP;

        ParkedVehicle pv;
        pv.setData(no, jk, st, datang, pulang);

        daftar.tambahData(pv); // === PERUBAHAN
    }

    daftar.tampilData(); // === PERUBAHAN
    return 0;
}
